### PR TITLE
RelaxNG autocompletion without namespace

### PR DESCRIPTION
**Pull request by angelozerr**
Fixes https://github.com/redhat-developer/vscode-xml/issues/791

Signed-off-by: azerr <azerr@redhat.com>

### CODE REVIEW

**Comment by angelozerr** [comment]
To test this PR, create a rng file which defines some required attribute with namespaces (`xml:lang`, `lmx:type`, vx:type) like this:

```xml
<grammar xmlns="http://relaxng.org/ns/structure/1.0"
  datatypeLibrary="http://www.w3.org/2001/XMLSchema-datatypes"
  xmlns:lmx="https://github.com/eclipse/lemminx"
  xmlns:vx="https://github.com/redhat-developer/vscode-xml">
  <start>
    <ref name="rootelt" />
  </start>
  <define name="rootelt">
    <element name="rootelt">
      <attribute name="xml:lang">
        <choice>
          <value>en</value>
          <value>de</value>
        </choice>
      </attribute>
      <attribute name="lmx:type">
        <choice>
          <value>dtd</value>
          <value>xsd</value>
          <value>rnc</value>
          <value>rng</value>
        </choice>
      </attribute>
      <element name="child">
        <attribute name="vx:type">
          <choice>
            <value>java</value>
            <value>binary</value>
          </choice>
        </attribute>
       </element>
    </element>
  </define>
</grammar>
```

Create an XML like this:

`<?xml-model href="simple.rng" schematypens="http://relaxng.org/ns/structure/1.0"?>`

and try the follwowing demo:

PrefixDemo1

It should generate the proper prefix for attribute and generate the xmlns namespace if need.

The xmlns attribute is not generate if it exists.

Given this xml:

```xml
<?xml-model href="simple.rng" schematypens="http://relaxng.org/ns/structure/1.0"?>
<rootelt xmlns:lmx="https://github.com/eclipse/lemminx"
	xmlns:vx="https://github.com/redhat-developer/vscode-xml"
	xml:lang="en" lmx:type="dtd">
	
</rootelt>
```

Please note that the generate of xmlns doesn't work if you open just the attribute completion and if the namespace doesn't exists  in XML it will not generate a prefix. It will require some more work to support that (use additionaltextEdit).

**Comment by angelozerr** [comment]
> If I make a short name for a namespace in the root element by adding the relevant attribute (eg. xmlns:asdf="https://google.ca"), and I reference an .rng that doesn't declare a ns, then I get 2 completion items for each element in the .rng: one with the prefix and without the prefix.

I noticed this problem indeed but it is not linked to this PR (the problem exists in master), but if I remember we need that for some usecases (we have some tests for that).

Could you create an issue please to study this problem in detail and try to fix it. Thanks!

File: org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/extensions/contentmodel/utils/XMLGenerator.java [hunk 1]
```diff
@@ -235,15 +270,43 @@ private int generate(Collection<CMAttributeDeclaration> attributes, int level, i
...
+				xml.addAttribute(attrName, value, level, true);
 			} else {
-				xml.addSingleAttribute(attributeDeclaration.getName(), value, true);
+				xml.addSingleAttribute(attrName, value, true);
 			}
 		}
 		return snippetIndex;
 	}
 
+	private String getAttrName(CMAttributeDeclaration attributeDeclaration, Map<String, String> prefixes) {
```
[the remark this answers is not in the document]
Comment (angelozerr) [reply org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/extensions/contentmodel/utils/XMLGenerator.java]: fixed

File: org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/extensions/xsd/contentmodel/CMXSDElementDeclaration.java [hunk 2]
```diff
@@ -83,14 +83,19 @@ public CMXSDElementDeclaration(CMXSDDocument document, XSElementDeclaration elem
...
 	}
 
 	@Override
 	public String getNamespace() {
 		return elementDeclaration.getNamespace();
 	}
+	
+	@Override
+	public String getPrefix(String namespaceURI) {
+		return null;
```
[the remark this answers is not in the document]
Comment (angelozerr) [reply org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/extensions/xsd/contentmodel/CMXSDElementDeclaration.java:98]: I add a TODO

[the remark this answers is not in the document]
Comment (angelozerr) [reply org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/extensions/xsd/contentmodel/CMXSDElementDeclaration.java:98]: See https://github.com/eclipse/lemminx/issues/1333

**Comment by angelozerr** [comment]
> If I make a short name for a namespace in the root element by adding the relevant attribute (eg. `xmlns:asdf="https://google.ca"`), and I reference an `.rng` that doesn't declare a `ns`, then I get 2 completion items for each element in the `.rng`: one with the prefix and without the prefix.

See https://github.com/eclipse/lemminx/issues/1334

**Review by datho7561 — APPROVED** [review]
Works great. Thanks, Angelo!

**Pull request merged**